core/src/main/java/org/elasticsearch/index/translog/Translog.java

```diff
@@ -1241,14 +1241,49 @@ void trimUnreferencedReaders() {
             minReferencedGen = Math.min(lastCommittedTranslogFileGeneration, minReferencedGen);
             final long finalMinReferencedGen = minReferencedGen;
             List<TranslogReader> unreferenced = readers.stream().filter(r -> r.getGeneration() < finalMinReferencedGen).collect(Collectors.toList());
+
+            Throwable firstException = null;
             for (final TranslogReader unreferencedReader : unreferenced) {
+                
                 Path translogPath = unreferencedReader.path();
-                logger.trace("delete translog file - not referenced and not current anymore {}", translogPath);
-                IOUtils.closeWhileHandlingException(unreferencedReader);
-                IOUtils.deleteFilesIgnoringExceptions(translogPath,
-                        translogPath.resolveSibling(getCommitCheckpointFileName(unreferencedReader.getGeneration())));
+                Path translogCheckpointPath = translogPath.resolveSibling(getCommitCheckpointFileName(unreferencedReader.getGeneration()));
+                logger.trace("delete translog files - not referenced and not current anymore {} and {}", translogPath, translogCheckpointPath);
+
+                try {
+                    unreferencedReader.close();
+                } catch (Throwable t) {
+                    if (firstException == null) {
+                        firstException = t;
+                    } else {
+                        firstException.addSuppressed(t);
+                    }
+                }
+
+                try {
+                    Files.delete(translogPath);
+                } catch (Throwable t) {
+                    if (firstException == null) {
+                        firstException = t;
+                    } else {
+                        firstException.addSuppressed(t);
+                    }
+                }
+
+                try {
+                    Files.delete(translogCheckpointPath);
+                } catch (Throwable t) {
+                    if (firstException == null) {
+                        firstException = t;
+                    } else {
+                        firstException.addSuppressed(t);
+                    }
+                }
+
+                readers.remove(unreferencedReader);
             }
-            readers.removeAll(unreferenced);
+
+            // NOTE: does nothing if firstException is null:
+            IOUtils.reThrow(firstException);
```

Comment:
This this can be simpler with org.elasticsearch.ExceptionsHelper#rethrowAndSuppress  - we just need to keep a list of throwable and deal with them in the end.